### ISSUE TITLE
Downloading NuPgk instead of building Roslyn

### ISSUE THREAD

**Issue by jaredpar** [ashmind/SharpLab#362]
This is a follow up to this tweet about the toll building dotnet/roslyn takes on SharpLab. 

https://twitter.com/ashmind/status/1022780582982344704

I have a PR out now that will make CI builds of Roslyn upload the produced NuPkg files as a build artifact. 

https://github.com/dotnet/roslyn/pull/32072

Once this is merged it should fairly quickly make its way to all of the feature and demo branches. At which point they will be producing NuPkg files on every merge as well. That should cover all the cases I'm aware of where SharpLab is currently building our repository and can now switch to just downloading the NuPkg files.

These are being uploaded as a build artifact vs. a NuGet feed. But Azure has a [nice API](https://docs.microsoft.com/en-us/rest/api/azure/devops/build/artifacts/get?view=azure-devops-rest-4.1) for discovering and downloading these on a branch basis. Hence it should fit into how SharpLab manages branches. 

I was actually hoping to submit a PR here instead of filing an issue. But looking through the source code I can't quite work out how the building of Roslyn + uploading is managed here. I found the Build-All.ps1 scripts  but can't see where they're hooked into the build. Is it just manually managed today? If it's something an external can help contribute to I'd be happy to send a PR here. Just need a few pointers on where to get started.

**Comment by ashmind** [comment]
Awesome, thanks a lot!

Also your consideration on doing a PR is much appreciated! However in this case I want to revisit the whole approach, so I'm not sure about the proper architecture yet. If the packaged dlls are signed with the same public key, then I don't need to rewrite references in SharpLab/MirrorSharp, so I might just deploy SL updates, and let branch web sites self-update by downloading artifacts directly. 

Just for clarity, current build runs on my local (because I wasn't sure how much a good VM would cost), and it directly runs Build-All, then Publish-All. I should have documented it properly, well, maybe after I redo it.

**Comment by jaredpar** [comment]
@ashmind 

> If the packaged dlls are signed with the same public key, then I don't need to rewrite references in SharpLab/MirrorSharp, so I might just deploy SL updates, and let branch web sites self-update by downloading artifacts directly

The packages will all have the same public keys. The binaries will be different in a few ways though:

- The packages do have the same keys. They will not be fully strong name signed, only public signed. But that shouldn't be any different than what you're getting if you build them yourself today.
- They will *possibly* have a different version number than what you're doing today.

**Comment by jaredpar** [comment]
Small update here. The name of the artifact is now "Packages - PreRelease". PR just merged.

**Comment by ashmind** [comment]
@jaredpar Thanks! Assuming it's already uploaded, can you give me a link to any (random) package?

**Comment by ashmind** [comment]
@jaredpar Question -- it seems that your PR was merged into e.g. `features/readonly-members`.
(at least according to `git branch --contains fa2350980ef479da64e49ef92f171b7350694f1d`)

However I can only find builds for `master`?

1. master: [branchName=refs/heads/master](https://dnceng.visualstudio.com/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_apis/build/builds?$top=10&repositoryId=dotnet%2Froslyn&repositoryType=GitHub&branchName=refs%2Fheads%2Fmaster) -- many results
2. features/readonly-members: [branchName=refs/heads/features/readonly-members](https://dnceng.visualstudio.com/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_apis/build/builds?$top=10&repositoryId=dotnet%2Froslyn&repositoryType=GitHub&branchName=refs%2Fheads%2Ffeatures%2Freadonly-members) ? -- no results

**Comment by jaredpar** [comment]
@ashmind no idea. My instinct was the same as yours. I tried a few variations and couldn't find a good answer here. Really can't find any guidance on how to filter these REST APIs for any parameters that seem logical. Let me ask around.

**Comment by alrz** [comment]
@ashmind I think majority of features are already merged to master. Would it be possible to at least make master available in the meantime?

**Comment by ashmind** [comment]
@alrz As for this ticket, it's unlikely I would implement it until there are more branches available -- I'm not very keen to rewrite the current solution for one single branch. This leaves two opportunities: either a PR to fix current build -- happy to accept -- or me building `master` manually one time.

I'll look into building it manually, but I can't promise any timeline for it.

**Comment by jaredpar** [comment]
@ashmind 

> As for this ticket, it's unlikely I would implement it until there are more branches available

Presently there should be three branches of interest: master, features/readonly-init and demos/records. Obviously not a huge jump from two though. But it does represent all of the branches of interest at the moment and will continue to do so going forward.

The limited amount of branches though isn't an indictment against our features / demo model but rather just a statement of where we are in the product cycle. As we are close to shipping C# 8.0 all of our active feature work has moved into master. Once we look forward to 9.0 branches will start popping up again for all the new features.

> This leaves two options: either a PR to fix current build -- happy to accept -- or me building `master` manually one time.

I'd be happy to look into a PR here for the NuPkg approach. I was actually a bit eager to try out the Azure APIs here and have some fun with them. I'm still a bit unsure though how you see this getting implemented. If you have a rough idea in mind I can see if I can build off of that. 

As for building master again I'd hope that not be necessary even in the case where you didn't change approaches. Given the code just relies on the NuPkg artifacts couldn't you just download the latest from our master branch and upload them where you normally would post build?

**Comment by alrz** [comment]
>The limited amount of branches though isn't an indictment against our features / demo model but rather just a statement of where we are in the product cycle.

Exactly my thoughts. We could start by master for now, and just add branches later on. This won't change "how" we call APIs and the underlying mechanism to feed it to sharplab.

**Comment by jaredpar** [comment]
@ashmind I figured out the issues using the Azure Build APIs. The query by default will return all builds: pull request, batched CI, etc ... The builds that sharplab is interested in are those which are not-pull request. Essentially individualCI. Here is a query for master

https://dev.azure.com/dnceng/public/_apis/build/builds?api-version=5.0&definitions=15&reasonfilter=individualCI&$top=15&branchName=refs/heads/master

**Comment by ashmind** [comment]
@jaredpar Thanks! That makes sense. Will the new branches of C# 9 appear automatically or require explicit inclusion in some pipeline filter? I don't really trust manual processes -- it's easy to lose track when there are other concerns.

Also, I can't find `features/readonly-members`? 
https://dev.azure.com/dnceng/public/_apis/build/builds?api-version=5.0&definitions=15&reasonfilter=individualCI&$top=15&branchName=refs/heads/features/readonly-members

On the implementation, current process is this (`Build-All.ps1`):

1. Build SharpLab
2. Update Roslyn repository
3. List branches
4. Go through each branch that matches conditions
5. Get and save branch metadata from git
6. Check if branch was updated since last build -- if not then go to 8
7. Build branch
8. Rewrite SharpLab binaries to the correct Roslyn version and collect all NuGet dependencies (done by AssemblyResolver project)

I was thinking to start with the following process (keeping `Build-All.ps1`):

1. List Roslyn branches using GitHub API
3. Go through each branch that matches conditions
4. Get and save branch metadata from GitHub API
5. Download latest build from Azure
6. If no new builds, go to 8
7. Unpack into a folder
8. Copy SharpLab version (per branch) (if already copied, update), skipping unchanged files
9. Put/ensure Directory.Build.targets that overrides referenced Roslyn version and includes folder with unpacked Azure build as a NuGet source
10. Build SharpLab, which will force it to use these Roslyn NuGet packages

If done right, this should preserve the dll dates if nothing has changed, which will be a big timesaver.
Currently rewrite in step 8 messes some dates, which causes ftp uploader to always re-upload some dlls, which is slow.

In any case, I don't mind writing it, but since you want to try it out -- happy to accept a PR.
Current build does not work anyways.

**Comment by ashmind** [comment]
@jaredpar Pls let me know if you still want to do the PR, otherwise I don't mind to pick it myself soon(ish).

**Comment by jaredpar** [comment]
I don't think I'm going to be able to get to this anytime soon. 😦

**Comment by ashmind** [comment]
@jaredpar No problem, I'll take a look.

However it seems I want this https://github.com/dotnet/roslyn/blob/8b75a0dcf999e64dae4c1f53fa995c70d1a449f4/azure-pipelines.yml#L2-L10 to include
```
- features/*
- demos/*
```
same as the `pr` section.

Should I create a PR for this?

**Comment by jaredpar** [comment]
@ashmind indeed it should. I thought it had already. Guessing that's why we aren't seeing any builds out of the readonly member branch.

Please do create a PR for that.

**Comment by ashmind** [comment]
Thanks @jaredpar! Implemented initial update and all available branches are up to date now.
Still some optimizations to be done, but I'll split those to separate tickets.

**Comment by ashmind** [comment]
@jaredpar Is it possible to include build number into the version produced by the build (current is `3.1.0-ci`, but something like `3.1.0-ci-131847`)? Not sure if `dotnet restore` can notice the change, so currently I'm manually deleting the package from under `--packages` folder before restore (if changed).

**Comment by jaredpar** [comment]
@ashmind 

> Is it possible to include build number into the version produced by the build (current is 3.1.0-ci, but something like 3.1.0-ci-131847)

where did you want to see this number added? Sorry lots of places these show up. Not sure which you mean here.

> Not sure if dotnet restore can notice the change, so currently I'm manually deleting the package from under --packages folder before restore (if changed).

Why does `restore` come into play here though? These packages are not uploaded to any NuGet server. Are you hosting them locally in a folder?

**Comment by ashmind** [comment]
> where did you want to see this number added? Sorry lots of places these show up. Not sure which you mean here.

Sure, what I mean is `<version>` in `nuspec` inside of `nupkg` inside `Packages - PreRelease.zip` artifact.
E.g. in this one https://dnceng.visualstudio.com/9ee6d478-d288-47f7-aacc-f6e6d082ae6d/_apis/build/builds/132595/artifacts?artifactName=Packages%20-%20PreRelease&api-version=5.1-preview.5&%24format=zip. Currently it seems to be `3.1.0-ci.final` always.

I can find the place in Roslyn code and create a PR, but not very soon.

> Why does restore come into play here though? These packages are not uploaded to any NuGet server. Are you hosting them locally in a folder?

Yep. I didn't want to unpack nupkg manually so it seemed to be the easiest solution.
